### DISCUSSION TITLE
RM2 Installation notes for 3.8.2

### DISCUSSION THREAD

**Discussion by UniqueName992** [timower/rM2-stuff#32] [General]
Here's my notes on installing `rocket` and `rm2fb` on RM2 version 3.8.2.
## Experimental Software Warning
A few people including myself have ended up needing to fix a boot-loop remarkable trying to install on this version. I don't recommend proceeding unless you're comfortable fixing things when they go wrong.
> This is an unsupported and unofficial configuration relying on bleeding edge patches and some hacks to work and is not recommended. See https://toltec-dev.org/#install-toltec for the latest supported version and consider using codexctl to downgrade instead.
### 1. Install 3.8.2 using codexctl (https://github.com/Jayy001/codexctl/releases/tag/1703028363).
```bash
./codexctl.bin install 3.8.2.1965
```
Disable automatic updates in the xochitl system settings.
### 2. Install toletc. Either re-enable a previously installed version:
```bash
toltecctl reenable
```
> **Or use this patched bootstrap if installing toltec fresh.**
> The patch changes the bootstrap script to use the latest wget available from https://github.com/toltec-dev/bootstrap. If you wish to make your own instead, download the usual bootstrap from the instructions at https://toltec-dev.org/#install-toltec and modify lines 82 and 83 to  v1.21.1-1 Wget, which works on Remarkable above v3.6 without the segmentation fault issue. 
> ```bash
> rm bootstrap
> wget http://raw.githubusercontent.com/UniqueName992/rm2-bootstrap-temp/main/bootstrap
> echo "a94ddad2aae1b560f5ff168dc309d4c4e39a4723136010253c31cdf3c5c5faad  bootstrap" | sha256sum -c && bash bootstrap
> ```
If you run into issues, removing any previous versions of binaries or entware may help before installing the bootstrap from the above instructions:
```bash
rm -r ~/.local/bin
rm -r ~/.entware
```
### 3. Mask the service remarkable-fail. This prevents ending up in a boot-loop situation if your device reboots before all steps are complete.
```bash
systemctl mask remarkable-fail
```
### 4. Restart the SSH session, or source the .bashrc to load the toltec binaries onto path:
```bash
source ~/.bashrc
```
### 5. Download the fixed xochitl package.
Either download from the pull request (https://github.com/toltec-dev/toltec/actions/runs/6991087513?pr=718) and extract it, then copy to the tablet or use the following commands which use a pre-extracted version:
```bash
wget https://raw.githubusercontent.com/UniqueName992/rm2-bootstrap-temp/main/xochitl_0.0.0-17_rmall.ipk
opkg remove xochitl
opkg install xochitl_0.0.0-17_rmall.ipk
opkg flag hold xochitl
```
The hold command will prevent this version being updated automatically. This is recommended as updates may not support 3.8.2 yet.
### 6. Install rm2fb
First install the toltec opkg, which won't work on this version so this will error. This is needed so Toltec doesn't install the toltec rm2fb and default to it later (which does not support 3.8.2):
```bash
opkg install rm2fb-client
opkg flag hold rm2fb-client
```
Now install the version for 3.8.2:
```bash
wget https://github.com/timower/rM2-stuff/releases/download/v0.1.1/rm2fb.ipk
opkg install rm2fb.ipk
```
### 7. install rocket
```bash
wget https://github.com/timower/rM2-stuff/releases/download/v0.1.1/rocket.ipk
opkg install rocket.ipk
```
### 8. Enable rm2fb
```
systemctl enable rm2fb
```
### 9. Set Rocket as default
```bash
systemctl disable xochitl
systemctl enable rocket
```
### 10. reboot
```bash
reboot now
```
All being well, Rocket should load on startup and can be used to load xochitl or any other installed packages.
### 11. Lastly, unmask the failure service:
```bash
systemctl unmask remarkable-fail
```
## Other Packages
At this point you should be fine to install other packages from toltec, most of which seem to work as expected showing up in rocket:
```bash
opkg install nao
opkg install koreader
```
The other packages from this rM2-stuff github can be installed either from toltec or similarly using wget.
## RM-Hacks
The new [Rm-hacks for 3.8.2](https://github.com/mb1986/rm-hacks) can also be installed as normal and work alongside toltec.
For the Lamy plugin with Rm-Hacks, a bit of extra work is needed to load the plugin by creating the following file:
```bash
nano /opt/etc/xochitl.env.d/lamy.env
```
with contents
```bash
export QT_QPA_GENERIC_PLUGINS=evdevlamy
```
## Launchers tested
* Rocket - Working well
* Remux - High CPU / battery drain Issue https://github.com/rmkit-dev/rmkit/issues/210
* Oxide - Not working

Thanks to @gullradriel for testing

**Comment by data-flux** [comment]
Good stuff! I don't suppose there's a way to use another launcher than rocket? I'm not a big fan of its sleep screen.

**Reply by UniqueName992** [reply]
Yeah, I've just tested Remux and it seemed to get working well with the necessary mods. Just keep the `remarkable-fail` service masked off as you experiment to reduce risk of ending up in a boot-loop.
```bash
systemctl mask remarkable-fail
```

Looks like the symlinks are needed to the older paths, at least for Remux:
```bash
ln -s /opt/lib/librm2fb_client.so /usr/lib/librm2fb_client.so
ln -s /opt/lib/librm2fb_client.so.1 /usr/lib/librm2fb_client.so.1
ln -s /opt/lib/librm2fb_client.so.1.1.0 /usr/lib/librm2fb_client.so.1.1.0
```

And modifying the service file with the preload `Environment=LD_PRELOAD=/opt/lib/librm2fb_client.so`
`/lib/systemd/system/remux.service`:
```bash
[Unit]
Description=App launcher that supports multi-tasking applications
After=xochitl.service opt.mount
StartLimitInterval=30
StartLimitBurst=5
Conflicts=draft.service tarnish.service

[Service]
ExecStart=/opt/bin/remux
Restart=on-failure
RestartSec=5
Environment="HOME=/home/root"
Environment=LD_PRELOAD=/opt/lib/librm2fb_client.so

[Install]
WantedBy=multi-user.target
```

**Reply by gullradriel** [reply]
@UniqueName992 I tested with Remux and Rocket and it worked, but not Oxide.
BTW I used ```Environment="LD_PRELOAD=/opt/lib/librm2fb_client.so"```

**Reply by UniqueName992** [reply]
Thanks, same results here testing Oxide (Tarnish)
I'll add a not to the main post

**Reply by gullradriel** [reply]
One more note:
remux works, but eats 40% of the CPU after a while. It's not eating the CPU at the start, but after a while.
Didn't investigated it more than that, but I tried to switch launchers a few times, and each time remux ended up eating 40% of the CPU constantly.
Only rocket seems to be going nice, without eating up the machine.
I ended up using rocket.

**Reply by data-flux** [reply]
Thanks @gullradriel, I was wondering why my battery was draining more quickly in the last few days.

**Reply by timower** [reply]
Interesting, I wonder if it's just a bug in remux or if it's my rm2fb + remux that's the issue.
For the different paths of `rm2fb_client.so`, https://github.com/timower/rM2-stuff/commit/e915fcbeb34a4ea1f452b1bc97f045eb5d402de6 should help

**Reply by UniqueName992** [reply]
Re. Remux Paths 
On validating, it's enough to pass the LD_PRELOAD alone or make the symlinks alone. Remux can launch with either configuration (and the high cpu usage issue exists with both configurations).

Here's a screenshot of the Remux High CPU issue, it seems to happen after a couple of switches and a notebook is opened in Xochitl, and remains high once the notebook is closed or xochitl is switched away from. Killing Xochitl reduces the CPU usage again. The high CPU usage is reported to belong to Remux.
![image](https://github.com/timower/rM2-stuff/assets/84731173/332515ac-6ee2-4082-b4cf-99ac05c282a5)

**Comment by UniqueName992** [comment]
Interesting edge case, after experimenting Rocket duplicated my icons on the main screen.
It looks like a package (possibly Draft) made a symlink from `/etc/draft/` to `/opt/etc/draft`. In this instance, Rocket read both directories and presents them as duplicate icons. Very interesting I can then have multiple instances of the same program open.

**Comment by Cyrus-Harding** [comment]
I had to use a new version of codexctl, and I had to use
```bash
wget http://toltec-dev.org/bootstrap
echo "72f08437001523c1eba86dabc054e9fe81b3fb12532f9eb41db0e7e29ff47f72  bootstrap" | sha256sum -c && bash bootstrap --force
```
